sessionctx/stmtctx/stmtctx.go

```diff
@@ -730,9 +730,7 @@ func (sc *StatementContext) SetMessage(msg string) {
 func (sc *StatementContext) GetWarnings() []SQLWarn {
 	sc.mu.Lock()
 	defer sc.mu.Unlock()
-	warns := make([]SQLWarn, len(sc.mu.warnings))
-	copy(warns, sc.mu.warnings)
-	return warns
+	return sc.mu.warnings
```

Comment:
Will return warnings directly cause conflict?

Reply:
This code date back to very early commit:

```
5230f5eabd7 (Ewan Chou    2016-11-29 14:25:38 +0800 407) // GetWarnings gets warnings.
5230f5eabd7 (Ewan Chou    2016-11-29 14:25:38 +0800 408) func (sc *StatementContext) GetWarnings() []error {
5230f5eabd7 (Ewan Chou    2016-11-29 14:25:38 +0800 409)        sc.mu.Lock()
5230f5eabd7 (Ewan Chou    2016-11-29 14:25:38 +0800 410)        warns := make([]error, len(sc.mu.warnings))
5230f5eabd7 (Ewan Chou    2016-11-29 14:25:38 +0800 411)        copy(warns, sc.mu.warnings)
5230f5eabd7 (Ewan Chou    2016-11-29 14:25:38 +0800 412)        sc.mu.Unlock()
5230f5eabd7 (Ewan Chou    2016-11-29 14:25:38 +0800 413)        return warns
5230f5eabd7 (Ewan Chou    2016-11-29 14:25:38 +0800 414) }
```

**The commit is sooooooo old that it's hard to figure out why it's written like that.**

The change should work, if not, at least we can see where the data race happen and try to fix again.
@Defined2014

Reply:
without copy, does the lock necessary here?

Reply:
> without copy, does the lock necessary here?

I think so ... `GetWarnings / SetWarnings` may be visited by many different component, maybe concurrently.

Reply:
Got it. The value of `sc.mu.warnings` will be changed by SetWarnings.